x/tokenomics/keeper/settle_pending_claims.go

```diff
@@ -160,25 +161,49 @@ func (k Keeper) SettlePendingClaims(ctx sdk.Context) (
 // This is the height at which the proof window closes.
 // If the proof window closes and a proof IS NOT required -> settle the claim.
 // If the proof window closes and a proof IS required -> only settle it if a proof is available.
-func (k Keeper) getExpiringClaims(ctx sdk.Context) (expiringClaims []prooftypes.Claim) {
+func (k Keeper) getExpiringClaims(ctx sdk.Context) (expiringClaims []prooftypes.Claim, err error) {
 	blockHeight := ctx.BlockHeight()
 
-	// TODO_TECHDEBT: Optimize this by indexing claims appropriately
-	// and only retrieving the claims that need to be settled rather than all
-	// of them and iterating through them one by one.
-	claims := k.proofKeeper.GetAllClaims(ctx)
+	// NB: This error can be safely ignored as on-chain SharedQueryClient implementation cannot return an error.
+	sharedParams, _ := k.sharedQuerier.GetParams(ctx)
+	claimWindowSizeBlocks := sharedParams.GetClaimWindowCloseOffsetBlocks()
+	proofWindowSizeBlocks := sharedParams.GetProofWindowCloseOffsetBlocks()
+
+	previousSessionEndHeight := blockHeight -
+		int64(sharedParams.GetGracePeriodEndOffsetBlocks()+
+			claimWindowSizeBlocks+
+			proofWindowSizeBlocks+1)
+
+	allClaims := k.proofKeeper.GetAllClaims(ctx)
+	_ = allClaims
+
+	for {
+		claimsRes, err := k.proofKeeper.AllClaims(ctx, &prooftypes.QueryAllClaimsRequest{
+			Pagination: &query.PageRequest{
+				Offset: uint64(len(expiringClaims)),
+			},
+			Filter: &prooftypes.QueryAllClaimsRequest_SessionEndHeight{
+				SessionEndHeight: uint64(previousSessionEndHeight),
```

Comment:
Wouldn't it be easier to just use `SessionNumber` for filtering?

Reply:
I'm not sure but at the moment we don't have any such index; we have:

```protobuf
 oneof filter {
    string supplier_address= 2;
    string session_id = 3;
    uint64 session_end_height = 4;
  }
```

Even if we had a `session_number` index, we would still have to do something more sophistocated than simply `currentHeight / num_blocks_per_session`. Do you think it's worth adding a `TODO_CONSIDERATION` or something?

Reply:
Maybe.

My concern is about the potential case where the session we want to settle is n-2 or older (relative to the current session).

If claim and proof window offsets do not allow that, then we should be fine, Otherwise we should consider other means to retrieve the session we want to settle.

Reply:
This method will be called for each block, progressively "draining" all expiring claims as the current height climbs. Since this is part of consensus, all sessions considered here MUST be from the same session number.